Source/WebCore/platform/graphics/FontFeatureValues.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FontFeatureValues.h"
+
+#include <wtf/text/TextStream.h>
+
+namespace WebCore {
+
+void add(Hasher& hasher, const FontFeatureValues& fontFeatureValues)
+{
+    auto hashTags = [&hasher](const Tags& tags) {
+        add(hasher, tags.isEmpty());
+        for (const auto& tag : tags)
+            add(hasher, tag.key, tag.value);
+    };
+    hashTags(fontFeatureValues.styleset);
+    hashTags(fontFeatureValues.stylistic);
+    hashTags(fontFeatureValues.characterVariant);
+    hashTags(fontFeatureValues.swash);
+    hashTags(fontFeatureValues.ornaments);
+    hashTags(fontFeatureValues.annotation);
+}
+
+void FontFeatureValues::upsert(const Ref<FontFeatureValues> toUpsert)
+{
+    // ALWAYS_LOG_WITH_STREAM(stream << "upsert " << toUpsert << " into " << *this);
+    auto upsertTags = [](auto& into, const Tags& tags) {
+        for (const auto& tag : tags)
+            into.set(tag.key, tag.value);
+    };
+    upsertTags(styleset, toUpsert->styleset);
+    upsertTags(stylistic, toUpsert->stylistic);
+    upsertTags(characterVariant, toUpsert->characterVariant);
+    upsertTags(swash, toUpsert->swash);
+    upsertTags(ornaments, toUpsert->ornaments);
+    upsertTags(annotation, toUpsert->annotation);
```

Comment:
I've never heard of `upsert` until now, it's in fact never used in the WebKit codebase aside from the code you've added. I think the usual terminology here is just `set`.

Reply:
It's a fairly common term: https://wiki.postgresql.org/wiki/UPSERT https://en.wiktionary.org/wiki/upsert
It conveys more meaning than "set"

Reply:
Seems common in the database world, not really in WebKit.
In fact, our own HashMap in WTF uses `set` as terminology

Reply:
(Even JS Map() uses `set`)

Reply:
But it's not a `set`. 

If I have `this` object with tags `[foo=>1; bar=>2]` and the `other` object with tags `[foo=>5, lol=>10]`.

`this.set(B)` would result in `this` being exactly `[foo=>5; lol=>10]`.
`this.upsert(B)` would result in `this` being [foo=>5; bar=>2; lol=>10]`

Reply:
`updateOrInsert()` seems more explicit then.

Reply:
My concern is mostly the lack of usage in WebKit: https://searchfox.org/wubkat/search?q=upsert&path=&case=false&regexp=false

Reply:
As Tim noted, I think `updateOrInsert` is likely the preferred name in WebKit, per style guidelines that prefer [full words over abbreviations](https://webkit.org/code-style-guidelines/#names) or (in this case) portmanteaus. I'm not sure this counts as one of the cases where the abbreviation is easier to understand, given that there's no other usage in WebKit, and it doesn't reflect the name of a DOM API (like `tabIndex` in the style guidelines example).

Perhaps, if we want to keep a short name, we could consider `replace`, or `merge`?

Reply:
I might be biased by my background in Postgres world 🙃 I will follow your opinions and change it to `updateOrInsert`.